rootchain/samuel/samuel.go

```diff
@@ -0,0 +1,580 @@
+	default:
+		return nil, errUnknownPayloadType
+	}
+}
+
+// startEventLoop starts the SAMUEL event monitoring loop, which retrieves
+// events from the Event Tracker, bundles them, and sends them off to other nodes
+func (s *SAMUEL) startEventLoop() {
+	subscription := s.eventTracker.Subscribe()
+
+	go func() {
+		for ev := range subscription {
+			// Get the raw event data as bytes
+			data, err := ev.Marshal()
+			if err != nil {
+				s.logger.Warn(fmt.Sprintf("unable to marshal Event Tracker event, %v", err))
+
+				continue
+			}
+
+			// Get the hash and the signature of the event
+			hash := crypto.Keccak256(data)
+			signature, blockNum, err := s.signer.Sign(hash)
+
+			if err != nil {
+				s.logger.Warn(fmt.Sprintf("unable to sign Event Tracker event, %v", err))
+
+				continue
+			}
+
+			// Push the SAM to the local SAMP
+			sam := rootchain.SAM{
+				Hash:          types.BytesToHash(hash),
+				Signature:     signature,
+				ChildBlockNum: blockNum,
+				Event:         ev,
+			}
+
+			if err := s.samp.AddMessage(sam); err != nil {
+				s.logger.Warn(fmt.Sprintf("unable to add event with hash %s to SAMP, %v", sam.Hash, err))
+
+				continue
+			}
+
+			// Publish the signature for other nodes
+			if err := s.transport.Publish(sam.ToProto()); err != nil {
+				s.logger.Warn(
+					fmt.Sprintf("unable to publish SAM message with hash %s to SAMP, %v", sam.Hash, err),
+				)
+
+				continue
+			}
+		}
+	}()
+}
+
+// Stop stops the SAMUEL module and any underlying modules
+func (s *SAMUEL) Stop() error {
+	// Stop the Event Tracker
+	if err := s.eventTracker.Stop(); err != nil {
+		return fmt.Errorf(
+			"unable to gracefully stop event tracker, %w",
+			err,
+		)
+	}
+
+	return nil
+}
+
+// SaveProgress notifies the SAMUEL module of which events
+// are committed to the blockchain
+func (s *SAMUEL) SaveProgress(
+	contractAddr types.Address, // local Smart Contract address
+	input []byte, // method with argument data
+) {
+	if contractAddr != types.StringToAddress(s.eventData.localAddress.String()) {
+		s.logger.Warn(
+			fmt.Sprintf("Attempted to save progress for unknown contract %s", contractAddr),
+		)
+
+		return
+	}
+
+	// Decode the inputs
+	methodID := s.eventData.methodABI.ID()
+	params, err := s.eventData.methodABI.Inputs.Decode(
+		input[len(methodID):],
+	)
+
+	if err != nil {
+		s.logger.Error(
+			fmt.Sprintf("Unable to decode event params for contract %s, %v", contractAddr, err),
+		)
+
+		return
+	}
+
+	// Make sure it's of a correct type
+	castParams, castErr := params.(map[string]interface{})
+	if err != nil {
+		s.logger.Error(
+			fmt.Sprintf("Unable to properly cast input params, %v", castErr),
+		)
+
+		return
+	}
+
+	switch s.eventData.payloadType {
+	case rootchain.ValidatorSetPayloadType:
+		// The method needs to contain
+		// (validatorSet[], index, blockNumber)
+		index, _ := castParams["index"].(uint64)
+		blockNumber, _ := castParams["blockNumber"].(uint64)
```

Comment:
Why not check the second returns?

Reply:
We can assume the cast is correct - if the input params are not valid, they will not get decoded correctly